File: multiplication_trainer.go
// multiplication_trainer.go - Тренажёр таблицы умножения на Go (CLI)
package main

import (
    "bufio"
    "fmt"
    "math/rand"
    "os"
    "strconv"
    "strings"
    "time"
)

func getDifficulty() int {
    fmt.Println("Выберите сложность:")
    fmt.Println("1. Лёгкая (1-5)")
    fmt.Println("2. Средняя (1-10)")
    fmt.Println("3. Сложная (1-20)")
    reader := bufio.NewReader(os.Stdin)
    for {
        fmt.Print("Ваш выбор (1-3): ")
        input, _ := reader.ReadString('\n')
        input = strings.TrimSpace(input)
        choice, err := strconv.Atoi(input)
        if err == nil && choice >= 1 && choice <= 3 {
            switch choice {
            case 1:
                return 5
            case 2:
                return 10
            case 3:
                return 20
            }
        }
        fmt.Println("Пожалуйста, введите 1, 2 или 3.")
    }
}

func getQuestionCount() int {
    reader := bufio.NewReader(os.Stdin)
    for {
        fmt.Print("Сколько примеров вы хотите решить? (1-50): ")
        input, _ := reader.ReadString('\n')
        input = strings.TrimSpace(input)
        count, err := strconv.Atoi(input)
        if err == nil && count >= 1 && count <= 50 {
            return count
        }
        fmt.Println("Пожалуйста, введите число от 1 до 50.")
    }
}

func runTrainer(maxNum, totalQuestions int) {
    reader := bufio.NewReader(os.Stdin)
    correct := 0
    start := time.Now()
    for i := 1; i <= totalQuestions; i++ {
        a := rand.Intn(maxNum) + 1
        b := rand.Intn(maxNum) + 1
        fmt.Printf("\nПример %d/%d: %d x %d = ? ", i, totalQuestions, a, b)
        input, _ := reader.ReadString('\n')
        input = strings.TrimSpace(input)
        userAnswer, err := strconv.Atoi(input)
        correctAnswer := a * b
        if err == nil && userAnswer == correctAnswer {
            fmt.Println("✅ Правильно!")
            correct++
        } else {
            fmt.Printf("❌ Неправильно! Правильный ответ: %d\n", correctAnswer)
        }
    }
    elapsed := time.Since(start).Seconds()
    percent := float64(correct) / float64(totalQuestions) * 100
    fmt.Println("\n" + strings.Repeat("=", 30))
    fmt.Println("📊 РЕЗУЛЬТАТЫ")
    fmt.Printf("Всего примеров: %d\n", totalQuestions)
    fmt.Printf("Правильных ответов: %d\n", correct)
    fmt.Printf("Неправильных ответов: %d\n", totalQuestions-correct)
    fmt.Printf("Процент правильных: %.1f%%\n", percent)
    fmt.Printf("Время: %.2f сек.\n", elapsed)
    fmt.Println(strings.Repeat("=", 30))
}

func main() {
    rand.Seed(time.Now().UnixNano())
    fmt.Println("🧮 Тренажёр таблицы умножения")
    maxNum := getDifficulty()
    total := getQuestionCount()
    runTrainer(maxNum, total)
    fmt.Print("Хотите повторить? (y/n): ")
    reader := bufio.NewReader(os.Stdin)
    again, _ := reader.ReadString('\n')
    again = strings.TrimSpace(strings.ToLower(again))
    if again == "y" {
        main()
    } else {
        fmt.Println("До свидания!")
    }
}

File: multiplication_trainer.py
#!/usr/bin/env python3
"""
multiplication_trainer.py - Тренажёр таблицы умножения на Python (CLI)
"""
import random
import time

def get_difficulty():
    print("Выберите сложность:")
    print("1. Лёгкая (1-5)")
    print("2. Средняя (1-10)")
    print("3. Сложная (1-20)")
    while True:
        try:
            choice = int(input("Ваш выбор (1-3): "))
            if choice == 1:
                return 5
            elif choice == 2:
                return 10
            elif choice == 3:
                return 20
            else:
                print("Пожалуйста, введите 1, 2 или 3.")
        except ValueError:
            print("Введите число.")

def get_question_count():
    while True:
        try:
            count = int(input("Сколько примеров вы хотите решить? (1-50): "))
            if 1 <= count <= 50:
                return count
            print("Пожалуйста, введите число от 1 до 50.")
        except ValueError:
            print("Введите число.")

def run_trainer(max_num, total_questions):
    correct = 0
    start_time = time.time()
    for i in range(1, total_questions + 1):
        a = random.randint(1, max_num)
        b = random.randint(1, max_num)
        print(f"\nПример {i}/{total_questions}: {a} x {b} = ? ", end="")
        try:
            user_answer = int(input())
        except ValueError:
            print("❌ Введите число.")
            user_answer = -1
        correct_answer = a * b
        if user_answer == correct_answer:
            print("✅ Правильно!")
            correct += 1
        else:
            print(f"❌ Неправильно! Правильный ответ: {correct_answer}")
    end_time = time.time()
    elapsed = end_time - start_time
    percent = (correct / total_questions) * 100
    print("\n" + "="*30)
    print("📊 РЕЗУЛЬТАТЫ")
    print(f"Всего примеров: {total_questions}")
    print(f"Правильных ответов: {correct}")
    print(f"Неправильных ответов: {total_questions - correct}")
    print(f"Процент правильных: {percent:.1f}%")
    print(f"Время: {elapsed:.2f} сек.")
    print("="*30)

def main():
    print("🧮 Тренажёр таблицы умножения")
    max_num = get_difficulty()
    total = get_question_count()
    run_trainer(max_num, total)
    again = input("Хотите повторить? (y/n): ").strip().lower()
    if again == 'y':
        main()
    else:
        print("До свидания!")

if __name__ == "__main__":
    main()
